source/extensions/filters/http/cache/http_cache.cc

```diff
@@ -74,30 +73,63 @@ LookupRequest::LookupRequest(const Http::RequestHeaderMap& request_headers, Syst
 size_t stableHashKey(const Key& key) { return MessageUtil::hash(key); }
 size_t localHashKey(const Key& key) { return stableHashKey(key); }
 
-// Returns true if response_headers is fresh.
-bool LookupRequest::isFresh(const Http::ResponseHeaderMap& response_headers) const {
-  if (!response_headers.Date()) {
-    return false;
+bool LookupRequest::requiresValidation(const Http::ResponseHeaderMap& response_headers) const {
+  // TODO(yosrym93): Store parsed response cache-control in cache instead of parsing it on every
+  // lookup
+  absl::string_view cache_control =
+      response_headers.getInlineValue(response_cache_control_handle.handle());
+  const ResponseCacheControl response_cache_control(cache_control);
+
+  SystemTime response_time = CacheHeadersUtils::httpTime(response_headers.Date());
+  SystemTime::duration response_age = timestamp_ - response_time;
+
+  ASSERT(response_age >= std::chrono::seconds(0), "Response time is in the future.");
```

Comment:
Do a "<" comparison on timestamp_ vs response_time before doing the subtraction. And you are error-checking with an ASSERT. I'm OK with an assert but also return some kind of error status as well unless you can guarantee through code inspection this cannot occur.

The other alternative is a RELEASE_ASSERT. What is not OK is to fall past that in production code.